packages/react/src/navigation-menu/root/NavigationMenuRoot.tsx

```diff
@@ -0,0 +1,292 @@
+      prevTriggerElementRef.current = undefined;
+    }
+
+    setMounted(false);
+    onOpenChangeComplete?.(false);
+    setActivationDirection(null);
+    setFloatingRootContext(undefined);
+
+    currentContentRef.current = null;
+    closeReasonRef.current = undefined;
+  });
+
+  useOpenChangeComplete({
+    enabled: !actionsRef,
+    open,
+    ref: { current: popupElement },
+    onComplete() {
+      if (!open) {
+        handleUnmount();
+      }
+    },
+  });
+
+  const contextValue: NavigationMenuRootContext = React.useMemo(
+    () => ({
+      open,
+      value,
+      setValue,
+      mounted,
+      transitionStatus,
+      positionerElement,
+      setPositionerElement,
+      popupElement,
+      setPopupElement,
+      viewportElement,
+      setViewportElement,
+      activationDirection,
+      setActivationDirection,
+      floatingRootContext,
+      setFloatingRootContext,
+      currentContentRef,
+      nested,
+      rootRef,
+      beforeInsideRef,
+      afterInsideRef,
+      beforeOutsideRef,
+      afterOutsideRef,
+      prevTriggerElementRef,
+      delay,
+      closeDelay,
+      orientation,
+      listRef,
+    }),
+    [
+      open,
+      value,
+      setValue,
+      mounted,
+      transitionStatus,
+      positionerElement,
+      popupElement,
+      viewportElement,
+      activationDirection,
+      floatingRootContext,
+      nested,
+      delay,
+      closeDelay,
+      orientation,
+    ],
+  );
+
+  const jsx = (
+    <NavigationMenuRootContext.Provider value={contextValue}>
+      <TreeContext componentProps={componentProps} forwardedRef={forwardedRef}>
+        {componentProps.children}
+      </TreeContext>
+    </NavigationMenuRootContext.Provider>
+  );
+
+  if (!nested) {
+    // FloatingTree provides context to nested menus
+    return <FloatingTree>{jsx}</FloatingTree>;
+  }
+
+  return jsx;
+});
+
+function TreeContext(props: {
+  componentProps: NavigationMenuRoot.Props;
+  forwardedRef: React.ForwardedRef<HTMLElement>;
+  children: React.ReactNode;
+}) {
+  const {
+    className,
+    render,
+    defaultValue,
+    value: valueParam,
+    onValueChange,
+    actionsRef,
+    delay,
+    closeDelay,
+    orientation,
+    onOpenChangeComplete,
+    ...elementProps
+  } = props.componentProps;
+
+  const nodeId = useFloatingNodeId();
+  const { rootRef, nested } = useNavigationMenuRootContext();
+
+  const { open } = useNavigationMenuRootContext();
+
+  const state: NavigationMenuRoot.State = React.useMemo(
+    () => ({
+      open,
+    }),
+    [open],
+  );
+
+  const element = useRenderElement(nested ? 'div' : 'nav', props.componentProps, {
+    state,
+    ref: [props.forwardedRef, rootRef],
+    props: [{ 'aria-orientation': orientation }, elementProps],
+  });
+
+  return (
+    <NavigationMenuTreeContext.Provider value={nodeId}>
+      {element}
+    </NavigationMenuTreeContext.Provider>
+  );
+}
+
+export namespace NavigationMenuRoot {
+  export interface State {
```

Comment:
`nested` should also be a part of State so they can render different components based on this (as we do by default).